### PR TITLE
Add first draft for the ROS2 launch files generator

### PR DESCRIPTION
**Pull request by ipa-nhg**
This PR implements a first draft version of the launch file generator for ROS2. This version allows from the following model:

```
RosSystem { Name 'move_base_experiment' 
RosComponents ( 
    ComponentInterface { name base_driver  FromRosNode "ridgeback.base_driver.base_driver"
        RosPublishers{
            RosPublisher 'joint_states' { RefPublisher "ridgeback.base_driver.base_driver.joint_states"},
            RosPublisher odometry { RefPublisher "ridgeback.base_driver.base_driver.odometry"} , RosPublisher sondar_scan { RefPublisher "ridgeback.base_driver.base_driver.sondar_scan" } } RosSubscribers {
            RosSubscriber vel_command { RefSubscriber "ridgeback.base_driver.base_driver.cmd_vel"}}},
    ComponentInterface { name scan_driver FromRosNode "hokuyo_node.hokuyo_node.hokuyo_node"
        RosPublishers{
            RosPublisher 'scan' { RefPublisher 'hokuyo_node.hokuyo_node.hokuyo_node.scan'},
            RosPublisher 'diagnostics' { RefPublisher 'hokuyo_node.hokuyo_node.hokuyo_node.diagnostics'} , RosPublisher "hokuyo/parameter_descriptions" { RefPublisher "hokuyo_node.hokuyo_node.hokuyo_node.hokuyo/parameter_descriptions" } , RosPublisher "hokuyo/parameter_updates" { RefPublisher "hokuyo_node.hokuyo_node.hokuyo_node.hokuyo/parameter_updates" } } RosSrvClients { RosServiceClient "hokuyo/set_parameters" { RefClient "hokuyo_node.hokuyo_node.hokuyo_node.hokuyo/set_parameters" } , RosServiceClient "hokuyo/self_test" { RefClient "hokuyo_node.hokuyo_node.hokuyo_node.hokuyo/self_test" } } } , 
    ComponentInterface { name move_base FromRosNode "move_base.move_base.move_base" 
    	RosPublishers { 
    		RosPublisher cmd_vel { RefPublisher "move_base.move_base.move_base.cmd_vel" }}
    	RosSubscribers { 
    		RosSubscriber nav_goal { RefSubscriber "move_base.move_base.move_base.move_base/goal" }}} ,
    ComponentInterface { name  navigation FromRosNode "navigation.navigation.navigation"
    	RosPublishers { 
    		RosPublisher nav_goal { RefPublisher "navigation.navigation.navigation.move_base/goal" }}
    	RosSubscribers { 
    		RosSubscriber scan { RefSubscriber "navigation.navigation.navigation.scan" } , 
    		RosSubscriber odometry { RefSubscriber "navigation.navigation.navigation.odometry" } , 
    		 RosSubscriber joint_states { RefSubscriber "navigation.navigation.navigation.joint_states" } , 
    		RosSubscriber map { RefSubscriber "navigation.navigation.navigation.map" }
    	, RosSubscriber imu { RefSubscriber "navigation.navigation.navigation.imu" } } } )
    }} 
```
the auto-generation of the ROS package: https://github.com/ipa-nhg/test_ros2_code_extractor/tree/LaunchFileGenerator/move_base_experiment_ros2 
@ipa-hsd could you please check the generated code and let me know all the errors I introduced? 

Missed:
- [x] Remaps of interfaces
- [x] Definition of parameters
- [ ] Trigger of the ROS1 or ROS2 code generator. Currently both packages are generated always that the rossystem file is correct. From the point of view of the eclipse IDE we could create 2 different developer perspectives, one for ROS1 and other for ROS2 . But  I would like to evaluate how this can be done with the web interface (theia) before we change the implementation
- [ ] Extension of the rossystem model to support further ROS2 launch concepts. Depending on this extension I will consider to set at model level an attribute that specify the ROS version, this will simplify also the generation process

### CODE REVIEW

**Comment by ipa-hsd** [comment]
Node can be initialzed like:
```
 node_robot_state_publisher = Node(
        package='robot_state_publisher',
        executable='robot_state_publisher',
        name='robot_state_publisher',
        namespace=name_space,
    )
```
since you have executable and also the node name, that can be included in initialization I guess: https://github.com/ipa-nhg/test_ros2_code_extractor/blob/LaunchFileGenerator/move_base_experiment_ros2/launch/move_base_experiment.launch#L8-L9

**Comment by ipa-nhg** [comment]
@ipa-hsd thanks for the review.

I made the changes you suggested:
- for the generator code: https://github.com/ipa320/ros-model/pull/131/commits/9c05c5b67013ef850c8e62565bce03b5fc97cadc
- the resulted model the looks: https://github.com/ipa-nhg/test_ros2_code_extractor/commit/f29bded2e9197aa127fb1d1bbfdb52d6bb08407e

**Comment by ipa-hsd** [comment]
rebased from master after https://github.com/ipa320/ros-model/pull/136

**Comment by ipa-hsd** [comment]
Parameters are now generated for launch files
```
joy = Node(
    package="joy",
    executable="joy_node",
    name="joy",
    parameters=[
      { "port" : 1001},
      { "array_test" : ["Hello", "world", "Harsh"]},
      { "struct_test" : [
        { "first_element" : 5 }, 
        { "second_element" : ["ROS", "world"] }]},
      { "ip" : "Hello"}
    ]
  )
```
but note that structs are treating like lists. This needs to be tested.

**Comment by ipa-hsd** [comment]
After testing the launch file with nested parameters (struct types), it seems that nested parameters (at least in the above form) are not supported. Either the parameters have to be expanded to its resolved name
```
{ "struct_test/first_element" : 5 }, 
{ "struct_test/second_element" : ["ROS", "world"] }
```
or load them from a YAML file.

**Comment by ipa-hsd** [comment]
As per https://github.com/ipa320/ros-model/pull/131/commits/48096c20731c2d1c54ee771774cd7897c915ce09, the parameters generated for structs would be of the type:
```
parameters=[
      { "port" : 1001 },
      { "array_test" : ["Hello", "world", "Harsh"] },
      { "struct_test/first_element" : 3 },
      { "struct_test/second_element" : ["ROS", "world"] },
      { "ip" : "Hellooo" }
    ]
```
Also generating some additional files and fixing install of certain files to be able to compile the package and run the launch file.

**Comment by ipa-nhg** [comment]
I have done a quick test with a sample system and it seems to work without any problem ✔️ 

I will try it with a more complex system and then fix the tests (https://github.com/ipa320/ros-model/pull/131/checks?check_run_id=2729737648) before we merge it. Thanks a lot @ipa-hsd

**Comment by ipa-hsd** [comment]
@ipa-nhg remap is still pending before we merge.

> I will try it with a more complex system
I doubt it will work for 2 layers of nesting, at least the param name might fail is my guess.

**Comment by ipa-hsd** [comment]
Just leaving it here: https://docs.ros.org/en/foxy/Guides/Launch-files-migration-guide.html   
XML format is supported for ROS2 and can be considered for sprint 2.

**Comment by ipa-nhg** [comment]
@ipa-hsd Thanks I added the link to the general issue: https://github.com/ipa320/ros-model/issues/109#issue-713727043

**Comment by ipa-nhg** [comment]
@ipa-hsd I fixed the ROS1 generator, could you please try it out? https://github.com/ipa320/ros-model/pull/131/commits/34fa197b04fb540b17ed000d585c327ff61817dd

**Comment by ipa-nhg** [comment]
I tested the ROS2 generator, it doesn't give errors, then we can merge it.

But a couple of things are missed:
- it is not considering the "forced" connections (TopicConnections, ServiceConnections and ActionConnections)
- the namespaces are not properly checked , if my input model is:

```
RosSystem { Name 'My'  RosComponents ( 
     ComponentInterface { name scan1 NameSpace test_ns FromRosNode "cob_sick_s300.cob_sick_s300.cob_sick_s300" 
      RosPublishers { RosPublisher scan { RefPublisher "cob_sick_s300.cob_sick_s300.cob_sick_s300.scan" } , 
      RosPublisher diagnostics { RefPublisher "cob_sick_s300.cob_sick_s300.cob_sick_s300.diagnostics" } } } ) }
```

This generates the following launch file:

```
from launch import LaunchDescription
from launch_ros.actions import Node

def generate_launch_description():
  ld = LaunchDescription()

  scan1 = Node(
    package="cob_sick_s300",
    executable="cob_sick_s300",
    name="scan1",
    namespace="test_ns"
    ,
    remappings=[
      ("scan", "scan"),
      ("diagnostics", "diagnostics")
    ]
  )

  ld.add_action(scan1)

  return ld

```
but the generated remaps are not adding changes and the interfaces shouldn't be remapped.

As we pretended to have a draft for ROS2, the proposed changes are OK for now. We will review the ROS2 generators for a future release.

**Comment by ipa-hsd** [comment]
> but the generated remaps are not adding changes and the interfaces shouldn't be remapped.

Yes I got the same results for ROS1 as well. the comparison for e.g.:
```
if (!rosPublisher.name.equals(compile_topic_name(rosPublisher.publisher,component.check_ns))) { // }
```
is not working

**Comment by ipa-nhg** [comment]
I think this is the correct version:

https://github.com/ipa320/ros-model/pull/131/commits/34fa197b04fb540b17ed000d585c327ff61817dd#diff-5aeece87e4ded60798c42b8de4e4fcd32b2ac3b368d175d628719a15f01c2d16R140

**Comment by ipa-hsd** [comment]
Fixed in https://github.com/ipa320/ros-model/pull/131/commits/a7f8854cf19f05fb86fa075e2b559e9443517536

File: plugins/de.fraunhofer.ipa.rossystem.xtext/src/de/fraunhofer/ipa/rossystem/generator/LaunchFileCompiler_ROS2.xtend [hunk 1]
```diff
@@ -78,6 +78,67 @@ def generate_launch_description():
...
+	}
+
+	def String compile_remappings_str(ComponentInterface component) {
+		var remap_str = "";
+		var count = component.rospublisher.length
+					+ component.rossubscriber.length
+					+ component.rosserviceserver.length
+					+ component.rosserviceclient.length
+					+ component.rosactionserver.length
+					+ component.rosactionclient.length;
```
Comment (ipa-hsd) [inline plugins/de.fraunhofer.ipa.rossystem.xtext/src/de/fraunhofer/ipa/rossystem/generator/LaunchFileCompiler_ROS2.xtend:92]: just realized this calculation does not make sense.

Comment (ipa-nhg) [reply plugins/de.fraunhofer.ipa.rossystem.xtext/src/de/fraunhofer/ipa/rossystem/generator/LaunchFileCompiler_ROS2.xtend:92]: is an open comma producing an error? if yes, you can create a boolean flag "HasRemaps" set to true if you enter one of the if conditions, then you can add the comma to all the remap_str modifications and at the end remove the last character.

I mean:


```
HasRemaps=false;

    for (rosPublisher : component.rospublisher) {
			if (!((prefix(NS)+rosPublisher.name).equals(compile_topic_name(rosPublisher.publisher, NS)))) {
				remap_str += "\t(\"" + rosPublisher.publisher.name + "\", \"" + rosPublisher.name + "\"),/n";
        HasRemaps=true;
			}
		}
		for (rosSubscriber : component.rossubscriber) {
			if (!((prefix(NS)+rosSubscriber.name).equals(compile_topic_name(rosSubscriber.subscriber, NS)))) {
				remap_str += "\t(\"" + rosSubscriber.subscriber.name + "\", \"" + rosSubscriber.name + "\"),/n";
				HasRemaps=true;
			}
		}

.....



if (HasRemaps){
  remap_str=remap_str.substring(0,remap_str.lenght-2)
  remap_str +="]"
}

```

**Pull request merged**